### PR TITLE
Adds the option to choose the function of encoding/decoding cookies

### PR DESCRIPTION
**Pull request by jorgeaguirre**
Principally adds the option to choose the function of encoding/decoding cookies. Additionally removes unnecessary operations while parsing cookies, since cookie attributes are being ignored.

### CODE REVIEW

**Comment by tomas** [comment]
Hmm. Why do we need different functions for encoding/decoding cookies?

**Comment by jorgeaguirre** [comment]
Because not all backends are well implemented, and in some cases cookies that were encoded with èncodeURIComponent` are not recognized, thus becomes necessary the use another encoding function.

**Comment by jorgeaguirre** [comment]
I will put it in another way. For some applications `needle` is used to make requests to third party servers. The way that cookies are now being encoded makes impossible to be properly authenticated by some of them. 
The problem reside in the encoding function, because `encodeURIComponent` escapes characters like `=` , `:` and in some cases the receiving server may not properly decode these characters. So if we add the option to choose different encoding function the problem goes away.

**Comment by tomas** [comment]
Can you share an example/spec/snippet/whatever to illustrate your point?

**Comment by jorgeaguirre** [comment]
OK, in this example you can see that the server not recognise the `JSESSIONID`cookie and set it again,

```javascript
var needle = require('needle');

var url = 'http://civil.poderjudicial.cl/CIVILPORWEB/';

var opts = {
  'follow_set_cookies': true
};

needle.get(url, opts, function(error, response) {
  opts.cookies = response.cookies;

  needle.post(url + 'AtPublicoDAction.do', null, opts, function(error, response) {
    console.log('(Request) Cookies header:')
    console.log(response.req._headers.cookie);
    console.log('\n(Response) Set-Cookie header:')
    console.log(response.cookies);
  });
});
```

but if the encoding function is changed , the server recognise the `JSESSIONID` cookie.

```javascript
var needle = require('needle');

var url = 'http://civil.poderjudicial.cl/CIVILPORWEB/';

var opts = {
  'follow_set_cookies': true,
  'cookieEncoder': encodeURI
};

needle.get(url, opts, function(error, response) {
  opts.cookies = response.cookies;

  needle.post(url + 'AtPublicoDAction.do', null, opts, function(error, response) {
    console.log('(Request) Cookies header:')
    console.log(response.req._headers.cookie);
    console.log('\n(Response) Set-Cookie header:')
    console.log(response.cookies);
  });
});
```

BTW what is the correct way to use `follow_set_cookie` properly?

**Comment by tomas** [comment]
`follow_set_cookies` requires the `follow_max` (or `follow`) option to be set to a number higher than 0. Otherwise no redirects are followed and thus, no cookies are set in the process.

I'll take a look at this in greater detail when I have a minute.

**Comment by Narzerus** [comment]
+1 on this, I'm having the same problem. @tomas any estimate on when you could decide wether or not to merge this PR?

Thanks  @jorgeaguirre for finding this issue, I was going crazy trying to figure out what was wrong with my requests

**Comment by tomas** [comment]
Hey @Narzerus, 

I haven't had a minute so I can't give a clear answer. I first need to perform a few tests in order to see the gravity of the issue. Thing is, I want Needle to "just work" so it shouldn't be necessary to have the use choose the cookie encoder -- i.e. Needle should be smart enough to determine which one to use and provide the data correctly.

Also there are no specs, and this is not a light PR!

**Comment by Narzerus** [comment]
@tomas agreed, I think the proper approach is to avoid needle's user to have to configure and re-configure for their specific usage.

Let me know how this goes, I don't want to have to fork this library to roll my own solution as I'd be missing future features and fixes.

Thanks for taking the time btw

**Comment by tomas** [comment]
@jorgeaguirre What do you think? Is there any way we can make the detection work automagically?

**Comment by jorgeaguirre** [comment]
we can use instead  of `encodeURIComponent` a function that encode only forbidden characters, based in the newest standard  [RFC6265](http://tools.ietf.org/html/rfc6265#section-4.1.1). 
In this is specified than allowed characters are:
````
0x21: !
0x23-2B: #$%&'()*+
0x2D-3A: -./0123456789: 
0x3C-5B: <=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[
0x5D-7E: ]^_`abcdefghijklmnopqrstuvwxyz{|}~
````
Or more generally `US-ASCII` characters excluding `CTLs`, `whitespace`, `DQUOTE`, `comma`, `semicolon`, and `backslash`.

this can be donde using like this

````javascript
// example of write to header function
var EXCLUDED_CHARS = /[\"\\;\s\"\,]/g;
function write(cookies) {
return Object.keys(cookies).map(function(key){
    var encodedKey = key.replace(EXCLUDED_CHARS, encodeURIComponent);
    var encodedValue = cookies[key].replace(EXCLUDED_CHARS, encodeURIComponent);
    return encodedKey + '=' +  encodedValue;
}).join('; ');
}

````

**Comment by jorgeaguirre** [comment]
What you think? I can make a PR with a encoding/decoding function that is RFC 6265 compliant

**Comment by tomas** [comment]
That'd be cool -- as long as we can have a unified way of handling cookies that doesn't require any guessing from the user. :)

**Comment by jorgeaguirre** [comment]
I will add cookie specs tomorrow, to test that everything is working right.

**Comment by Narzerus** [comment]
@jorgeaguirre thanks for your work dude, much appreciated

**Comment by jorgeaguirre** [comment]
I think with this is enough to test that encoding is working right. I don't add tests for `follow_set_cookies` feature.

File: test/cookies_spec.js [hunk 1]
```diff
@@ -1,75 +1,174 @@
...
-      })
+      needle.get('http://allcookies.test', opts, function (error, response) {
+        var cookieString = response.req._headers.cookie;
 
-    })
+        cookieString.should.be.type('string');
 
-  })
+        cookieString.split(/\s*;\s*/).forEach(function (pair) {
+          COOKIE_PAIR.test(pair).should.be.exactly(true);
```
Comment (tomas) [inline test/cookies_spec.js]: Shoudn't we test for the value themselves rather than just testing against the regex?

File: test/cookies_spec.js [hunk 2]
```diff
@@ -1,75 +1,174 @@
...
+        });
+        done();
+      });
+    });
 
-  describe('when passing an object on options.cookies', function() {
+    it('dont have to encode allowed characters', function (done) {
+      var COOKIE_PAIR = /^([^=\s]+)\s*=\s*("?)\s*(.*)\s*\2\s*$/,
+        KEY_INDEX = 1,
+        VALUE_INEX = 3;
```
Comment (tomas) [inline test/cookies_spec.js]: Is this always the case? (key at 1, value at 3)

**Comment by tomas** [comment]
Bravo, looks beautiful. I mentioned a thing or two up there, but aside from that it's perfect.

File: test/cookies_spec.js (hunk 2 above)
Comment (jorgeaguirre) [inline test/cookies_spec.js]: yes

File: test/cookies_spec.js (hunk 1 above)
Comment (jorgeaguirre) [inline test/cookies_spec.js]: Yes. But  in this test i want to validate that is a valid cookie string. I can add a test to compare the cookie string with a known value.

File: test/cookies_spec.js (hunk 2 above)
Comment (jorgeaguirre) [inline test/cookies_spec.js]: the second index could be a double quote or nothing, eliminating them from the value.

File: test/cookies_spec.js [hunk 3]
```diff
@@ -1,75 +1,174 @@
...
-  })
-
-  before(function(done) {
-    server.listen(function() {
-      port = this.address().port;
-      done();
+var needle = require('../'),
+  sinon = require('sinon'),
+  should = require('should'),
+  nock = require('nock'),
```
Comment (Narzerus) [inline test/cookies_spec.js]: Great HTTP mocking library

File: test/cookies_spec.js (hunk 1 above)
Comment (tomas) [inline test/cookies_spec.js]: That'd be good.

**Comment by Narzerus** [comment]
Please have a look at #118, @jorgeaguirre did you notice this while refactoring `lib/cookies.js`?

File: lib/cookies.js [hunk 4]
```diff
@@ -1,77 +1,81 @@
...
+function getFirstPair(str) {
+  var index = str.indexOf('\x3B');
+  return index === -1 ? str : str.substr(0, index);
+}
+
+// Private functions
+
+// Returns a encoded copy of str based on RFC6265 S4.1.1.
+function encodeCookieComponent(str) {
+  return str.replace(EXCLUDED_CHARS, encodeURIComponent);
```
Comment (Narzerus) [inline lib/cookies.js]: This line fails if str is a number, you should to a `toString()` before replacing

**Comment by Narzerus** [comment]
@jorgeaguirre regarding my previous comment on multiple cookies, it looks like the bug is still present in your PR.

File: lib/cookies.js (hunk 4 above)
Comment (jorgeaguirre) [inline lib/cookies.js]: That it's a good idea, i will add this and add exceptions.

**Comment by Narzerus** [comment]
Any news on this?

**Comment by tomas** [comment]
Will take a look in a minute. @jorgeaguirre anything's that missing?

**Comment by Narzerus** [comment]
Nice, will be waiting

**Comment by tomas** [comment]
I just reverted this PR after seeing that for some reason the cookies_spec.js breaks almost all the other tests. Would you mind fixing it @jorgeaguirre?

**Comment by jorgeaguirre** [comment]
Sorry i was too lazy to create my own ssl keys to run all tests. The problem its related with [nock](https://github.com/pgte/nock#enabledisable-real-http-request) since it disable real http request.

I run all test now and i see the problem, i need to figure out how to encapsulate the use of `nock` just for `cookies_spec.js`. I will fix this soon.

**Comment by tomas** [comment]
Ok, let me know when you're done. I think @Narzerus is getting impatient, haha.

**Comment by Narzerus** [comment]
Hahah dont mean to put pressure on you guys, we're just running a project
which is currently being heavily affected by the issue. I'm truly grateful
for your work guys <3

**Comment by ash0080** [comment]
plz reopen it

provide a practical case here:

```
// set in cookies option
cookies: {
    userCookieKey: 'in%2B8ekX8rcEkxWkA10jAiw%3D%3D'
 },

// what real request is:
 headers: 
   { Accept: '*/*',
     'Content-Length': '0',
     Cookie: 'userCookieKey=in%252B8ekX8rcEkxWkA10jAiw%253D%253D',
     Host: 'httpbin.org',
     'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 10_3_1 like Mac OS X) AppleWebKit/603.1.30 (KHTML, like Gecko) Version/10.0 Mobile/14E304 Safari/602.1',
     'X-Amzn-Trace-Id': 'Root=1-5e8956ef-a847ab877d0b978935daf9ec',
     'X-Requested-With': 'XMLHttpRequest' },
  json: null,
  method: 'POST',
  origin: '114.93.35.129',
  url: 'http://httpbin.org/anything' }

```

**Comment by tomas** [comment]
@ash0080 Please provide a PR with a fix

**Comment by ash0080** [comment]
@tomas when I get some time, On top of that, plz review the last PR first

**Pull request merged**